File: client/assets/scripts/utils/AESCipher.ts
import forge from 'node-forge';

/**
 * 与 server/internal/helper 一致：
 * - 内层 AES-GCM + PBKDF2(100000, SHA-256, 32B)
 * - 外层字节包：packed[2*i]=random[i], packed[2*i+1]=inner[i]；密钥材料 MD5(UTF8(deviceId)||random)
 * - JSON 传输：StdBase64(packed)
 */
export class AESCipher {
    private static fillRandomBytes(out: Uint8Array): void {
        if (typeof globalThis !== 'undefined' && globalThis.crypto?.getRandomValues) {
            globalThis.crypto.getRandomValues(out);
        } else {
            for (let i = 0; i < out.length; i++) out[i] = Math.floor(Math.random() * 256);
        }
    }

    private static toBase64(bytes: Uint8Array): string {
        const binary = String.fromCharCode.apply(null, Array.from(bytes));
        return btoa(binary);
    }

    private static fromBase64(base64: string): Uint8Array {
        const binary = atob(base64);
        const bytes = new Uint8Array(binary.length);
        for (let i = 0; i < binary.length; i++) {
            bytes[i] = binary.charCodeAt(i);
        }
        return bytes;
    }

    private static uint8ArrayToForgeBytes(bytes: Uint8Array): string {
        let binary = '';
        for (let i = 0; i < bytes.length; i++) {
            binary += String.fromCharCode(bytes[i]);
        }
        return binary;
    }

    private static forgeBytesToUint8Array(binary: string): Uint8Array {
        const bytes = new Uint8Array(binary.length);
        for (let i = 0; i < binary.length; i++) {
            bytes[i] = binary.charCodeAt(i);
        }
        return bytes;
    }

    /** 内层 [salt16][iv12][cipher+tag]，与 EncryptToBytes 一致 */
    private static encryptInner(data: string, keyMaterial: string): Uint8Array {
        const salt = new Uint8Array(16);
        const iv = new Uint8Array(12);
        this.fillRandomBytes(salt);
        this.fillRandomBytes(iv);

        const keyBytes = forge.util.createBuffer(keyMaterial, 'utf8').getBytes();
        const saltBytes = this.uint8ArrayToForgeBytes(salt);
        const derivedKeyBytes = forge.pkcs5.pbkdf2(
            keyBytes,
            saltBytes,
            100000,
            32,
            forge.md.sha256.create()
        );

        const ivBytes = this.uint8ArrayToForgeBytes(iv);
        const cipher = forge.cipher.createCipher('AES-GCM', derivedKeyBytes);
        cipher.start({
            iv: ivBytes,
            additionalData: '',
            tagLength: 128,
        });
        cipher.update(forge.util.createBuffer(data, 'utf8'));
        cipher.finish();

        const ciphertextBytes = cipher.output.getBytes();
        const tagBytes = cipher.mode.tag.getBytes();
        const combinedCiphertext = ciphertextBytes + tagBytes;
        const combinedCiphertextUint8 = this.forgeBytesToUint8Array(combinedCiphertext);

        const result = new Uint8Array(salt.length + iv.length + combinedCiphertextUint8.length);
        result.set(salt, 0);
        result.set(iv, salt.length);
        result.set(combinedCiphertextUint8, salt.length + iv.length);
        return result;
    }

    private static decryptInner(encrypted: Uint8Array, keyMaterial: string): string {
        const salt = encrypted.slice(0, 16);
        const iv = encrypted.slice(16, 28);
        const combinedCiphertext = encrypted.slice(28);

        const keyBytes = forge.util.createBuffer(keyMaterial, 'utf8').getBytes();
        const saltBytes = this.uint8ArrayToForgeBytes(salt);
        const derivedKeyBytes = forge.pkcs5.pbkdf2(
            keyBytes,
            saltBytes,
            100000,
            32,
            forge.md.sha256.create()
        );

        const ivBytes = this.uint8ArrayToForgeBytes(iv);
        const combinedBytes = this.uint8ArrayToForgeBytes(combinedCiphertext);
        const tagLength = 16;
        const ciphertextBytes = combinedBytes.slice(0, -tagLength);
        const tagBytes = combinedBytes.slice(-tagLength);

        const decipher = forge.cipher.createDecipher('AES-GCM', derivedKeyBytes);
        decipher.start({
            iv: ivBytes,
            additionalData: '',
            tag: forge.util.createBuffer(tagBytes),
            tagLength: 128,
        });
        decipher.update(forge.util.createBuffer(ciphertextBytes));
        if (!decipher.finish()) {
            return '';
        }
        return decipher.output.toString();
    }

    private static md5DevicePlusRandom(deviceId: string, random: Uint8Array): string {
        const dev = new TextEncoder().encode(deviceId);
        const combined = new Uint8Array(dev.length + random.length);
        combined.set(dev);
        combined.set(random, dev.length);
        const md = forge.md.md5.create();
        md.update(this.uint8ArrayToForgeBytes(combined));
        return md.digest().toHex();
    }

    private static mergeInterleaveBytes(random: Uint8Array, inner: Uint8Array): Uint8Array {
        const n = inner.length;
        if (random.length !== n) {
            throw new Error('interleave length mismatch');
        }
        const out = new Uint8Array(2 * n);
        for (let i = 0; i < n; i++) {
            out[2 * i] = random[i];
            out[2 * i + 1] = inner[i];
        }
        return out;
    }

    private static splitInterleaveBytes(packed: Uint8Array): { random: Uint8Array; inner: Uint8Array } {
        if (packed.length % 2 !== 0) {
            throw new Error('invalid packed length');
        }
        const n = packed.length / 2;
        const random = new Uint8Array(n);
        const inner = new Uint8Array(n);
        for (let i = 0; i < n; i++) {
            random[i] = packed[2 * i];
            inner[i] = packed[2 * i + 1];
        }
        return { random, inner };
    }

    /**
     * 返回字节交错包（长度 2*L），第二个参数为设备 ID（与请求头 X-Device-ID 一致）
     */
    public static encrypt(data: string, deviceId: string): Uint8Array {
        if (!deviceId) {
            throw new Error('empty device id');
        }
        const dummy = this.encryptInner(data, '0');
        const L = dummy.length;
        const random = new Uint8Array(L);
        this.fillRandomBytes(random);
        const keyMaterial = this.md5DevicePlusRandom(deviceId, random);
        const inner = this.encryptInner(data, keyMaterial);
        if (inner.length !== L) {
            throw new Error(`cipher length changed: ${inner.length} vs ${L}`);
        }
        return this.mergeInterleaveBytes(random, inner);
    }

    /**
     * 解 encrypt 产生的字节包
     */
    public static decrypt(packed: Uint8Array, deviceId: string): string {
        const { random, inner } = this.splitInterleaveBytes(packed);
        const keyMaterial = this.md5DevicePlusRandom(deviceId, random);
        return this.decryptInner(inner, keyMaterial);
    }

    /** StdBase64(encrypt(...))，供 JSON 字段 encrypt 使用 */
    public static encryptToBase64(data: string, deviceId: string): string {
        return this.toBase64(this.encrypt(data, deviceId));
    }

    public static decryptFromBase64(encrypted: string, deviceId: string): string {
        return this.decrypt(this.fromBase64(encrypted), deviceId);
    }
}

File: cypoker-server/internal/helper/aes-cipher.go
package helper

import (
	"crypto/aes"
	"crypto/cipher"
	"crypto/rand"
	"crypto/sha256"
	"encoding/base64"
	"errors"
	"fmt"
	"server/internal/utils"

	"golang.org/x/crypto/pbkdf2"
)

// 使用PBKDF2派生密钥
func getCryptoKey(keyMaterial string, salt []byte) ([]byte, error) {
	// PBKDF2派生密钥，与JS版本保持一致：
	// iterations: 100000, hash: SHA-256, key length: 32 bytes (256 bits)
	key := pbkdf2.Key([]byte(keyMaterial), salt, 100000, 32, sha256.New)
	return key, nil
}

// 将字节数组转换为Base64
func toBase64(bytes []byte) string {
	return base64.StdEncoding.EncodeToString(bytes)
}

// fromBase64 将Base64转换为字节数组
func fromBase64(base64Str string) ([]byte, error) {
	return base64.StdEncoding.DecodeString(base64Str)
}

// Encrypt 加密
// data 需要加密的 JSON 字符串
// key 密钥
// 返回加密后的字符串，base64编码
// 返回错误信息
func Encrypt(data string, key string) (string, error) {
	// 准备明文数据
	plaintext := []byte(data)

	// 生成随机salt (16字节) 和 iv (12字节，AES-GCM推荐)
	salt := make([]byte, 16)
	if _, err := rand.Read(salt); err != nil {
		return "", err
	}

	iv := make([]byte, 12)
	if _, err := rand.Read(iv); err != nil {
		return "", err
	}

	// 派生密钥
	cryptoKey, err := getCryptoKey(key, salt)
	if err != nil {
		return "", err
	}

	// 创建AES cipher
	block, err := aes.NewCipher(cryptoKey)
	if err != nil {
		return "", err
	}

	// 创建GCM模式
	gcm, err := cipher.NewGCM(block)
	if err != nil {
		return "", err
	}

	// 加密数据
	ciphertext := gcm.Seal(nil, iv, plaintext, nil)

	// 组装结果: [salt] + [iv] + [ciphertext]
	result := make([]byte, 0, len(salt)+len(iv)+len(ciphertext))
	result = append(result, salt...)
	result = append(result, iv...)
	result = append(result, ciphertext...)

	// 返回Base64
	return toBase64(result), nil
}

// Decrypt 解密
// encryptStr 加密字符串，base64编码
// key 密钥
// 返回解密后的 JSON 字符串
// 返回错误信息
func Decrypt(encrypted string, key string) (string, error) {
	// 将Base64转换为字节数组
	encryptedData, err := fromBase64(encrypted)
	if err != nil {
		return "", err
	}

	// 检查数据长度是否足够
	if len(encryptedData) < 28 { // 16(salt) + 12(iv)
		return "", errors.New("encrypted data too short")
	}

	// 解析salt和iv
	salt := encryptedData[:16]
	iv := encryptedData[16:28]
	ciphertext := encryptedData[28:]

	// 派生密钥
	cryptoKey, err := getCryptoKey(key, salt)
	if err != nil {
		return "", err
	}

	// 创建AES cipher
	block, err := aes.NewCipher(cryptoKey)
	if err != nil {
		return "", err
	}

	// 创建GCM模式
	gcm, err := cipher.NewGCM(block)
	if err != nil {
		return "", err
	}

	// 解密数据
	plaintext, err := gcm.Open(nil, iv, ciphertext, nil)
	if err != nil {
		return "", err
	}

	// 返回解码后的字符串
	return string(plaintext), nil
}

// EncryptToBytes 加密数据并返回字节数组
func EncryptToBytes(data string, key string) []byte {
	plaintext := []byte(data)

	// 生成随机salt (16字节) 和 iv (12字节)
	salt := make([]byte, 16)
	iv := make([]byte, 12)
	rand.Read(salt)
	rand.Read(iv)

	// 派生密钥
	cryptoKey, _ := getCryptoKey(key, salt)

	// 创建AES cipher
	block, _ := aes.NewCipher(cryptoKey)
	gcm, _ := cipher.NewGCM(block)

	// 加密数据
	ciphertext := gcm.Seal(nil, iv, plaintext, nil)

	// 组装结果: [salt] + [iv] + [ciphertext]
	result := make([]byte, 0, len(salt)+len(iv)+len(ciphertext))
	result = append(result, salt...)
	result = append(result, iv...)
	result = append(result, ciphertext...)

	return result
}

// DecryptBytes 解密字节数组
func DecryptBytes(encrypted []byte, key string) ([]byte, error) {
	// 检查数据长度
	if len(encrypted) < 28 {
		return nil, errors.New("encrypted data too short")
	}

	// 解析salt和iv
	salt := encrypted[:16]
	iv := encrypted[16:28]
	ciphertext := encrypted[28:]

	// 派生密钥
	cryptoKey, err := getCryptoKey(key, salt)
	if err != nil {
		return nil, err
	}

	// 创建AES cipher
	block, err := aes.NewCipher(cryptoKey)
	if err != nil {
		return nil, err
	}

	// 创建GCM模式
	gcm, err := cipher.NewGCM(block)
	if err != nil {
		return nil, err
	}

	// 解密数据
	plaintext, err := gcm.Open(nil, iv, ciphertext, nil)
	if err != nil {
		return nil, err
	}

	return plaintext, nil
}

// mergeInterleaveBytes 与 splitInterleaveBytes 互逆：out[2*i]=random[i], out[2*i+1]=inner[i]
func mergeInterleaveBytes(random, inner []byte) []byte {
	n := len(inner)
	out := make([]byte, 2*n)
	for i := 0; i < n; i++ {
		out[2*i] = random[i]
		out[2*i+1] = inner[i]
	}
	return out
}

func splitInterleaveBytes(packed []byte) (random, inner []byte, err error) {
	if len(packed)%2 != 0 {
		return nil, nil, errors.New("invalid packed length")
	}
	n := len(packed) / 2
	random = make([]byte, n)
	inner = make([]byte, n)
	for i := 0; i < n; i++ {
		random[i] = packed[2*i]
		inner[i] = packed[2*i+1]
	}
	return random, inner, nil
}

// EncryptPackedBytes 设备端字节包：长度 2*L，L 为内层 AES 包长；密钥材料 MD5(deviceID + string(random))
func EncryptPackedBytes(plaintext, deviceID string) ([]byte, error) {
	if deviceID == "" {
		return nil, errors.New("empty device id")
	}
	dummy := EncryptToBytes(plaintext, "0")
	L := len(dummy)
	random := make([]byte, L)
	if _, err := rand.Read(random); err != nil {
		return nil, err
	}
	key := utils.MD5(deviceID + string(random))
	inner := EncryptToBytes(plaintext, key)
	if len(inner) != L {
		return nil, fmt.Errorf("cipher length changed: %d vs %d", len(inner), L)
	}
	return mergeInterleaveBytes(random, inner), nil
}

// DecryptPackedBytes 解 EncryptPackedBytes 产出
func DecryptPackedBytes(packed []byte, deviceID string) (string, error) {
	random, inner, err := splitInterleaveBytes(packed)
	if err != nil {
		return "", err
	}
	key := utils.MD5(deviceID + string(random))
	pt, err := DecryptBytes(inner, key)
	if err != nil {
		return "", err
	}
	return string(pt), nil
}
